events/amqp_logger.go

```diff
@@ -31,40 +34,134 @@ func (AMQPEventLoggerFactory) NewEventHandler(config map[interface{}]interface{}
 }
 
 type amqpWriteCloser struct {
-	queue amqp.Queue
-	ch    *amqp.Channel
-	conn  *amqp.Connection
+	queue    amqp.Queue
+	ch       *amqp.Channel
+	conn     *amqp.Connection
+	config   *amqpConfig
+	ctx      context.Context
+	cancel   context.CancelFunc
+	messages chan []byte
 }
 
-func (wc amqpWriteCloser) Write(data []byte) (int, error) {
+func newAMQPWriteCloser(config *amqpConfig) amqpWriteCloser {
+	ctx, cancel := context.WithCancel(context.Background())
+	ret := amqpWriteCloser{
+		config:   config,
+		ctx:      ctx,
+		cancel:   cancel,
+		messages: make(chan []byte, config.bufferSize),
+	}
+	go func() {
+		ret.connect()
+		for {
+			select {
+			case m := <-ret.messages:
+				{
+					ret.sendMessage(m)
+				}
+			case <-ret.ctx.Done():
+				{
+					var retErr error
+					//TODO: Update to errors.Join on go 1.20 update
+					if ret.ch != nil {
+						if err := ret.ch.Close(); err != nil {
+							retErr = err
+						}
+					}
+					if ret.conn != nil {
+						if err := ret.conn.Close(); err != nil {
+							if retErr != nil {
+								retErr = fmt.Errorf("%v; %w", retErr, err)
+							} else {
+								retErr = err
+							}
+						}
+					}
+					if retErr != nil {
+						logrus.Errorf("error closing amqp connections: %v", retErr)
+						return
+					}
+					logrus.Info("closed connection to amqp server")
+				}
+			}
+		}
+	}()
+
+	return ret
+}
+
+func (wc *amqpWriteCloser) sendMessage(message []byte) {
+	select {
```

Comment:
Unlikely, but you could still blow out the stack if you recurse enough times. 

```
for {
   trySend
   if success return
   if !closed {
      log
      return
   }
   reconnect
}
```

Reply:
Good catch, better to be safe here